src/opencl_common.c

```diff
@@ -1268,6 +1268,26 @@ static char *include_source(const char *pathname, int sequential_id, const char
 		full_path = quote_str(full_path);
 	}
 
+#ifdef __CYGWIN__
+	/*
+	 * Workaround for OpenCL kernels not found when run from Windows PowerShell.
+	 * CygWin (at least in PowerShell) maps the main JtR folder to "/".
+	 */
+	const char *sandbox_path;
+
+	sandbox_path = realpath(".", NULL);
+
+	if (sandbox_path && sandbox_path[0] == '/' && !strcmp(full_path, "/run/opencl")) {
+		int main_folder_fd;
+
+		if ((main_folder_fd = open("/run/opencl", O_DIRECTORY)) > 0) {
+			fchdir(main_folder_fd);
+			close(main_folder_fd);
+		}
```

Comment:
This is just a weird way to `chdir()`. My suggestion to use an fd was for a different purpose - to save the previous working directory and then restore it.

Reply:
Claudio, have you tried using simply `chdir()` here? While it's supposed to work the same as above, I admit I'm not 100% sure both have them same magic in them.

Reply:
The original idea:
```
fd = open(".", O_DIRECTORY);
chdir(/run/opencl);
do-stuff();
fchdir(fd);
```

But, what is `.`? What is `/`? The main problem in my patch it that I do not trust these "ideas" inside CygWin.

I do **open** the directory "/run/opencl", so if I succeed, I do something. Otherwise, do nothing but pray.

In my mind, that is safer. Probably, chdir() will do the same thing, but is it going to open the directory before do something else? But wait, do we care? It that important?

Reply:
I'm now experimenting with what might be a generic solution on Linux. In my testing, not chdir'ing back to the previous current directory results in other path_expand'ed pathnames being wrong, so e.g. there's issue caching the binary kernel. I wonder why you didn't(?) run into such issues with your patch on Windows.

Reply:
BTW, there's also this comment:

```c
        // include source is thread safe.
        build_opts = include_source("$JOHN/opencl", sequential_id, opts);
```

My idea of saving/restoring the cwd might not be thread-safe. As I recall, we experimented with building kernels from an OpenMP parallel region for descrypt-opencl, but disabled it since?